File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/options/unit/DistanceUnit.java
package wangdaye.com.geometricweather.common.basic.models.options.unit;

import android.content.Context;

import wangdaye.com.geometricweather.R;

public enum DistanceUnit {

    KM("km", 0, 1f),
    M("m", 1, 1000f),
    MI("mi", 2, 0.6213f),
    NMI("nmi", 3, 0.5399f),
    FT("ft", 4, 3280.8398f);

    private final String unitId;
    private final int unitArrayIndex;
    private final float unitFactor; // actual distance = distance(km) * factor.

    DistanceUnit(String id, int arrayIndex, float factor) {
        unitId = id;
        unitArrayIndex = arrayIndex;
        unitFactor = factor;
    }

    public String getUnitId() {
        return unitId;
    }

    public float getDistance(float km) {
        return km * unitFactor;
    }

    public String getDistanceText(Context context, float km) {
        return UnitUtils.formatFloat(km * unitFactor, 2) + "\u202f" +  getAbbreviation(context);
    }

    public String getAbbreviation(Context context) {
        return context.getResources().getStringArray(R.array.distance_units)[unitArrayIndex];
    }

    public String getDistanceVoice(Context context, float km) {
        return UnitUtils.formatFloat(km * unitFactor, 2)
                + "\u202f" +  context.getResources().getStringArray(R.array.distance_unit_voices)[unitArrayIndex];
    }

    public static DistanceUnit getInstance(String value) {
        switch (value) {
            case "m":
                return M;

            case "mi":
                return MI;

            case "nmi":
                return NMI;

            case "ft":
                return FT;

            default:
                return KM;
        }
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/options/unit/DistanceUnit.kt
package wangdaye.com.geometricweather.common.basic.models.options.unit

import android.content.Context
import wangdaye.com.geometricweather.R
import wangdaye.com.geometricweather.common.basic.models.options._basic.UnitEnum
import wangdaye.com.geometricweather.common.basic.models.options._basic.Utils
import wangdaye.com.geometricweather.common.utils.DisplayUtils

// actual distance = distance(km) * factor.
enum class DistanceUnit(
    override val id: String,
    override val unitFactor: Float
): UnitEnum<Float> {

    KM("km", 1f),
    M("m", 1000f),
    MI("mi", 0.6213f),
    NMI("nmi", 0.5399f),
    FT("ft", 3280.8398f);

    companion object {

        @JvmStatic
        fun getInstance(
            value: String
        ) = when (value) {
            "m" -> M
            "mi" -> MI
            "nmi" -> NMI
            "ft" -> FT
            else -> KM
        }
    }

    override val valueArrayId = R.array.distance_unit_values
    override val nameArrayId = R.array.distance_units
    override val voiceArrayId = R.array.distance_unit_voices

    override fun getName(context: Context) = Utils.getName(context, this)

    override fun getVoice(context: Context) = Utils.getVoice(context, this)

    override fun getValueWithoutUnit(valueInDefaultUnit: Float) = valueInDefaultUnit * unitFactor

    override fun getValueInDefaultUnit(valueInCurrentUnit: Float) = valueInCurrentUnit / unitFactor

    override fun getValueTextWithoutUnit(
        valueInDefaultUnit: Float
    ) = Utils.getValueTextWithoutUnit(this, valueInDefaultUnit, 2)!!

    override fun getValueText(
        context: Context,
        valueInDefaultUnit: Float
    ) = getValueText(context, valueInDefaultUnit, DisplayUtils.isRtl(context))

    override fun getValueText(
        context: Context,
        valueInDefaultUnit: Float,
        rtl: Boolean
    ) = Utils.getValueText(
        context = context,
        enum = this,
        valueInDefaultUnit = valueInDefaultUnit,
        decimalNumber = 2,
        rtl = rtl
    )

    override fun getValueVoice(
        context: Context,
        valueInDefaultUnit: Float
    ) = getValueVoice(context, valueInDefaultUnit, DisplayUtils.isRtl(context))

    override fun getValueVoice(
        context: Context,
        valueInDefaultUnit: Float,
        rtl: Boolean
    ) = Utils.getVoiceText(
        context = context,
        enum = this,
        valueInDefaultUnit = valueInDefaultUnit,
        decimalNumber = 2,
        rtl = rtl
    )
}